blosc/blosc2.c

```diff
@@ -2241,129 +2253,203 @@ static void t_blosc_do_job(void *ctxt)
+            pthread_mutex_lock(&context->count_mutex);
+            context->thread_giveup_code = cbytes;
+            pthread_mutex_unlock(&context->count_mutex);
+            break;
+          }
         }
-        cbytes = blosc_c(thcontext, bsize, leftoverblock, 0,
-                          ebsize, src, nblock_ * blocksize, tmp2, tmp, tmp3);
-      }
-    }
-    else {
-      if (flags & BLOSC_MEMCPYED) {
-        /* We want to memcpy only */
-        fastcopy(dest + nblock_ * blocksize,
-                  src + BLOSC_MAX_OVERHEAD + nblock_ * blocksize, (unsigned int)bsize);
-        cbytes = (int32_t)bsize;
+
+        pthread_mutex_lock(&context->count_mutex);
+        context->thread_nblock++;
+        nblock_ = context->thread_nblock;
+        pthread_mutex_unlock(&context->count_mutex);
+        /* Update counter for this thread */
+        ntbytes += cbytes;
+
+      } /* closes while (nblock_) */
+
+      /* Sum up all the bytes decompressed */
+      if ((!compress || (flags & BLOSC_MEMCPYED)) && (context->thread_giveup_code > 0)) {
+        /* Update global counter for all threads (decompression only) */
+        pthread_mutex_lock(&context->count_mutex);
+        context->output_bytes += ntbytes;
+        pthread_mutex_unlock(&context->count_mutex);
       }
-      else {
-        cbytes = blosc_d(thcontext, bsize, leftoverblock,
-                          src + sw32_(bstarts + nblock_),
-                          dest, nblock_ * blocksize, tmp, tmp2);
+    } else {
+      if (compress && !(flags & BLOSC_MEMCPYED)) {
+        /* Compression always has to follow the block order */
+        pthread_mutex_lock(&context->count_mutex);
+        context->thread_nblock++;
+        nblock_ = context->thread_nblock;
+        pthread_mutex_unlock(&context->count_mutex);
+        tblock = nblocks;
+      } else {
+        /* Decompression can happen using any order.  We choose
+        sequential block order on each thread */
+
+        /* Blocks per thread */
+        tblocks = nblocks / context->nthreads;
+        leftover2 = nblocks % context->nthreads;
+        tblocks = (leftover2 > 0) ? tblocks + 1 : tblocks;
+
+        nblock_ = thcontext->tid * tblocks;
+        tblock = nblock_ + tblocks;
+        if (tblock > nblocks) {
+          tblock = nblocks;
+        }
       }
-    }
 
-    /* Check whether current thread has to giveup */
-    if (context->thread_giveup_code <= 0) {
-      break;
-    }
+      /* Loop over blocks */
+      leftoverblock = 0;
+      while ((nblock_ < tblock) &&
+             (context->thread_giveup_code > 0)) {
+        bsize = blocksize;
+        if (nblock_ == (nblocks - 1) && (leftover > 0)) {
+          bsize = leftover;
+          leftoverblock = 1;
+        }
+        if (compress) {
+          if (flags & BLOSC_MEMCPYED) {
+            /* We want to memcpy only */
+            fastcopy(dest + BLOSC_MAX_OVERHEAD + nblock_ * blocksize,
+                src + nblock_ * blocksize, (unsigned int)bsize);
```

Comment:
For arguments in continuation lines, the convention is to align them with the first parameter.